coder-sdk/env.go

```diff
@@ -75,3 +78,127 @@ func (c Client) DialWsep(ctx context.Context, env *Environment) (*websocket.Conn
 	}
 	return conn, nil
 }
+
+// CreateEnvironmentRequest is used to configure a new environment
+type CreateEnvironmentRequest struct {
+	Name     string   `json:"name"`
+	ImageID  string   `json:"image_id"`
+	ImageTag string   `json:"image_tag"`
+	CPUCores float32  `json:"cpu_cores"`
+	MemoryGB int      `json:"memory_gb"`
+	DiskGB   int      `json:"disk_gb"`
+	GPUs     int      `json:"gpus"`
+	Services []string `json:"services"`
+}
+
+// CreateEnvironment sends a request to create an environment.
+func (c Client) CreateEnvironment(ctx context.Context, orgID string, req CreateEnvironmentRequest) (Environment, error) {
+	var env Environment
+	err := c.requestBody(
+		ctx,
+		http.MethodPost, "/api/orgs/"+orgID+"/environments",
+		req,
+		&env,
+	)
+	return env, err
+}
+
+type envUpdate struct {
+	Type string `json:"type"`
+}
+
+// WaitForEnvironmentReady watches the environment update websocket and waits for the "done" message type before returning.
+func (c Client) WaitForEnvironmentReady(ctx context.Context, envID string) error {
+	u := c.copyURL()
+	if c.BaseURL.Scheme == "https" {
+		u.Scheme = "wss"
+	} else {
+		u.Scheme = "ws"
+	}
+	u.Path = "/api/environments/" + envID + "/watch-update"
+
+	conn, resp, err := websocket.Dial(ctx, u.String(),
+		&websocket.DialOptions{
+			HTTPHeader: map[string][]string{
+				"Cookie": {"session_token=" + c.Token},
+			},
+		},
+	)
+	if err != nil {
+		if resp != nil {
+			return bodyError(resp)
+		}
+		return err
+	}
+
+	for {
+		m := envUpdate{}
+		err = wsjson.Read(ctx, conn, &m)
+		if err != nil {
+			return fmt.Errorf("read ws json msg: %w", err)
+		}
+		if m.Type == "done" {
+			break
+		}
+	}
+
+	return nil
+}
+
+type stats struct {
+	ContainerStatus string `json:"container_status"`
+	StatError       string `json:"stat_error"`
+	Time            string `json:"time"`
+}
+
+// WatchEnvironmentStats watches the environment update websocket for a given duration.
+func (c Client) WatchEnvironmentStats(ctx context.Context, envID string, duration time.Duration) error {
+	u := c.copyURL()
+	if c.BaseURL.Scheme == "https" {
+		u.Scheme = "wss"
+	} else {
+		u.Scheme = "ws"
+	}
+	u.Path = "/api/environments/" + envID + "/watch-stats"
+
+	conn, resp, err := websocket.Dial(ctx, u.String(),
+		&websocket.DialOptions{
+			HTTPHeader: map[string][]string{
+				"Cookie": {"session_token=" + c.Token},
+			},
+		},
+	)
+	if err != nil {
+		if resp != nil {
+			return bodyError(resp)
+		}
+		return err
+	}
+
+	statsCtx, statsCancel := context.WithTimeout(ctx, duration)
+	defer statsCancel()
+
```

Comment:
I'm not sure about this one @cmoog 

I think if he passes `statsCtx` to the websocket read that gives us two paths to cancellation that could occur at two different times. 

I think a cancellation of the parent context would propogate to `statsCtx` but not the other way around since a time-out is not always indicative of cancellation of the parent ctx.

Reply:
Oh, interesting. The two paths of cancellation is definitely a valid point. Especially considering you'd want a much shorter timeout for the initial `Dial`

I'd say that makes me in favor of something similar to what Colin is suggesting where the WebSocket opener just returns some watcher structure that holds additional methods for listening that are bounded by a different context.